src/ALCProxy.Communication/IProxyClient.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. 
+// The .NET Foundation licenses this file to you under the MIT license. 
+// See the LICENSE file in the project root for more information. 
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Loader;
+
+namespace ALCProxy.Communication
+{
+    /// <summary>
+    /// This currently is designed to only work in-process
+    /// TODO: set up to allow for construction of out-of-proc proxies
+    /// </summary>
+    public abstract class ALCClient : IProxyClient
+    {
+        //Can't make this an IServerObject directly due to the type-loading barrier
+        protected object _server;
+        protected string _serverTypeName;
+        protected Type _intType;
+        protected MethodInfo _callMethod;
+        public ALCClient(Type interfaceType, string serverName)
+        {
+            _intType = interfaceType;
+            _serverTypeName = serverName;
+        }
+        private Type FindTypeInAssembly(string typeName, Assembly a)
+        {
+            //find the type we're looking for
+            Type t = a.GetType(typeName);
+            if (t == null)
+            {
+                t = a.GetType(a.GetName().Name + "." + typeName);
+                if (t == null)
+                {
+                    throw new TypeLoadException("Proxy creation exception: No valid type while searching for the given type");
+                }
+            }
+            return t;
+        }
+        /// <summary>
+        /// Creates the link between the client and the server, while also passing in all the information to the server for setup
+        /// </summary>
+        /// <param name="alc">The target AssemblyLoadContext</param>
+        /// <param name="typeName">Name of the proxied type</param>
+        /// <param name="assemblyPath">path of the assembly to the type</param>
+        /// <param name="genericTypes">any generics that we need the proxy to work with</param>
+        public void SetUpServer(AssemblyLoadContext alc, string typeName, string assemblyPath, object[] constructorParams, Type[] genericTypes)
+        {
+            Assembly a = alc.LoadFromAssemblyPath(assemblyPath);
+            //find the type we're going to proxy inside the loaded assembly
+            Type objType = FindTypeInAssembly(typeName, a);
+            //Load *this* (ALCProxy.Communication) assembly into the ALC so we can get the server into the ALC
+            Assembly aa = alc.LoadFromAssemblyPath(Assembly.GetAssembly(typeof(ServerDispatch<>)).CodeBase.Substring(8));
```

Comment:
I don't understand the need for this - this will cause the assembly to be loaded twice which incurs non-trivial cost to the runtime. Among other things we can't reuse Ready2Run images, so it will be JITed again.
Since we own the code in this assembly we can make it ALC aware and thus it should not matter where it comes from - meaning it should he shared.
Going forward, if this would ever become part of the framework it would be shared almost by default anyway.

Reply:
For now, I was building most of the implementation in such a way that we can make sure that no shared assemblies are required to get plugins working. That said, if we wanted to enforce that plugins need some reference to the main program, and thus this API, we could probably run with shared assemblies and prevent the extra load cost. Is that what you're thinking we should be doing with the API here to prevent the double-load?

Reply:
As noted later on, I think the proxies should not make any assumptions about the ALC and how it loads assemblies. They should always load "through" that ALC, but they should not assume in which ALC the assembly will actually get loaded.
There are benefits to sharing and ALCs should be free to make that choice.

Note that this does not limit unloadability - the proxies should have the clear isolation regardless of where the assemblies ends up loaded. And so if they do get loaded into a separate ALC, then they will get unloaded. If they don't, then we only added potentially unnecessary cost to calling to it, but nothing will break due to it.

I get why you made it this way - it is relatively easy to see that we isolate fully I guess. That said I would still make the proxy classes itself an exception to this rule. But I guess it doesn't matter much.

Reply:
That makes a lot of sense to consider, and yeah, performance would be a lot better using shared assemblies there. I'll add a comment on how we want to do this to the design doc, since the issue from the current implementation does seem pretty important.

Reply:
Sounds great!